4_setup_train_ff_topologies/setup_train_ff_topo.py

```diff
@@ -177,40 +235,97 @@ def prepare_to_train(
     # Get starting forcefield
     offxml = pathlib.Path(offxml)
     logger.info(f"Loading force field: {offxml.resolve()}")
-    starting_ff = ForceField(offxml)
 
     # Optional molecule validation
     if validate_molecules:
-        dataset = validate_molecular_dataset(dataset, allow_undefined_stereo=True)
+        dataset = validate_molecular_dataset(dataset)
 
     total_molecules = len(dataset)
     logger.info(f"Processing {total_molecules} molecules.")
 
     # Process molecules and create interchanges
-    logger.info("Creating interchanges...")
-    all_smiles = []
-    all_interchanges = []
-    failed_molecules = 0
-
-    for entry in tqdm(dataset, desc="Creating interchanges"):
-        smiles: str = ""
-        try:
-            smiles = entry["smiles"]  # type: ignore[index]
-            mol = Molecule.from_mapped_smiles(smiles, allow_undefined_stereo=True)
-            interchange = starting_ff.create_interchange(mol.to_topology())
+    all_smiles = [entry["smiles"] for entry in dataset]  # type: ignore[index]
+    if n_cpus == 1:
+        maybe_interchanges = [
+            smiles_to_interchange(x, str(offxml))
+            for x in tqdm(all_smiles, desc="Creating Interchanges")
+        ]
+    else:
+        if n_cpus is None:
+            n_cpus = os.cpu_count() or 1
+
+        logger.info(f"Found {n_cpus} workers for interchange creation")
+        logger.info("Processing molecules in parallel...")
+        logger.info(f"Total molecules: {len(all_smiles)}")
+
+        logger.info("Starting parallel processing...")
+
+        with ProcessPoolExecutor(max_workers=n_cpus) as executor:
+            with tqdm(total=len(all_smiles), desc="Creating Interchanges") as pbar:
+                maybe_interchanges = []
+                processed_count = 0
+                failed_count = 0
+
+                try:
+                    # Submit all tasks
+                    future_to_smiles = {
+                        executor.submit(
+                            smiles_to_interchange, smiles, str(offxml)
+                        ): smiles
+                        for smiles in all_smiles
+                    }
+
+                    # Process completed tasks as they finish
+                    from concurrent.futures import as_completed
+
+                    for future in as_completed(future_to_smiles):
+                        result = future.result()
+                        maybe_interchanges.append(result)
+                        processed_count += 1
+                        pbar.update(1)
```

Comment:
Similar question as before -- does `as_completed` return in order for the zipping on line 318? Given that this is a dict, maybe you could just create a new list `maybe_smiles = []` and append to that in this loop as well to ensure ordering?

Reply:
(Or even just create the list of `filtered_interchanges`, `filtered_smiles` here and only append if result is not None)

Reply:
I ended up switching to `executor.map()` to preserve the order.